caffe2/operators/onnxifi_op.cc

```diff
@@ -0,0 +1,157 @@
+    Workspace* ws,
+    ::ONNX_NAMESPACE::TensorProto* t) {
+  // Set name
+  t->set_name(name);
+  const Blob* blob = ws->GetBlob(name);
+  CAFFE_ENFORCE(blob, "Blob ", name, " doesn't exist");
+
+  // Set dims
+  const auto shape = GetTensorShapeOfBlob(blob);
+  for (const auto i : shape.dims()) {
+    t->add_dims(i);
+  }
+
+  // Set values
+  CAFFE_ENFORCE(
+      blob->template IsType<TensorCPU>(),
+      "Initialization blob ",
+      name,
+      " needs to be either TensorCPU or TensorCUDA");
+  CPUTensorToTensorProto(blob->template Get<TensorCPU>(), t);
+}
+}
+
+template <>
+void OnnxifiOp<float, CPUContext>::BuildInitializationList(
+    Workspace* ws,
+    ::ONNX_NAMESPACE::GraphProto* g,
+    std::unordered_set<std::string>* initialization_list) {
+  const std::vector<string>& ws_blobs = ws->Blobs();
+
+  for (const auto& s : ws_blobs) {
+    auto it = initialization_list->find(s);
+    if (it != initialization_list->end()) {
+      auto* init_tensor = g->add_initializer();
+      BlobToTensorProto(s, ws, init_tensor);
+      initialization_list->erase(it);
+    }
+  }
+  CAFFE_ENFORCE(
+      initialization_list->empty(), "Unfulfilled initialization list");
+  for (const auto& t : g->initializer()) {
+    VLOG(2) << "Initializer: " << t.name();
+  }
+}
+
+template <>
+bool OnnxifiOp<float, CPUContext>::RunOnDevice() {
+  for (unsigned i = 0U; i < InputSize(); ++i) {
+    const auto& input_tensor = Input(i);
+    const auto& tensor_dims = input_tensor.dims();
+    auto& tensor_descriptor = input_desc_.at(i);
+    tensor_descriptor.dataType = ONNXIFI_DATATYPE_FLOAT32;
+    tensor_descriptor.memoryType = ONNXIFI_MEMORY_TYPE_CPU;
+    tensor_descriptor.dimensions = tensor_dims.size();
+    input_shapes_.emplace_back();
+    auto& input_shape = input_shapes_.back();
+    for (unsigned j = 0U; j < tensor_descriptor.dimensions; ++j) {
+      input_shape.push_back(tensor_dims[j]);
+    }
+    tensor_descriptor.shape = input_shape.data();
+    tensor_descriptor.buffer = (onnxPointer)(input_tensor.data<float>());
+  }
+
+  for (unsigned i = 0U; i < OutputSize(); ++i) {
+    auto* output_tensor = Output(i);
+    std::vector<TIndex> tensor_dims;
+    SetOutputShape(i, &tensor_dims);
+    output_tensor->Resize(tensor_dims);
+    auto& tensor_descriptor = output_desc_.at(i);
+    tensor_descriptor.dataType = ONNXIFI_DATATYPE_FLOAT32;
+    tensor_descriptor.memoryType = ONNXIFI_MEMORY_TYPE_CPU;
+    tensor_descriptor.dimensions = tensor_dims.size();
+    tensor_descriptor.shape = new uint64_t[tensor_descriptor.dimensions];
+    output_shapes_.emplace_back();
+    auto& output_shape = output_shapes_.back();
+    for (unsigned j = 0U; j < tensor_descriptor.dimensions; ++j) {
+      output_shape.push_back(tensor_dims[j]);
+    }
+    tensor_descriptor.shape = output_shape.data();
+    tensor_descriptor.buffer =
+        (onnxPointer)(output_tensor->mutable_data<float>());
+  }
+
+  CAFFE_ENFORCE_EQ(
+      lib_->onnxSetGraphIO(
```

Comment:
Its okish for the first implementation, but later we'd need to make sure we call `onnxSetGraphIO` only if input shape or buffer locations have changed.

Reply:
Yes.